ecoli/processes/metabolism_redux.py

```diff
@@ -0,0 +1,499 @@
+                self.Se.at[met, exch_name] = 1
+
+            self.Se[exch_name + " rev"] = np.zeros(self.n_mets).astype(np.int8)
+            self.Se.at[met, exch_name + " rev"] = -1
+
+        self.S_orig = csr_matrix(np.array(self.Sd, dtype=np.float64))
+        self.S_exch = csr_matrix(np.array(self.Se, dtype=np.float64))
+
+        _, self.n_exch_rxns = self.S_exch.shape
+
+        self.secretion_idx = np.where(self.S_exch.sum(axis=0) == -1)[0]
+
+    def solve(self,
+              homeostatic_targets: Mapping[str, float],
+              maintenance_target: float,
+              kinetic_targets: Mapping[str, float],
+              binary_kinetic_targets: Mapping[str, float],
+              objective_weights: Mapping[str, float],
+              upper_flux_bound: float = 100,
+              **kwargs) -> FlowResult:
+
+        # objective targets: dict of dicts to floats, for now
+        ## binary kinetic
+        # starting fluxes: optional for now, could reintroduce
+        #
+
+        homeostatic_arr = [[self.mets.index(met), target] for met, target in homeostatic_targets.items()]
+        homeostatic_idx, homeostatic_target = np.array(homeostatic_arr, dtype=np.int64)[:, 0], np.array(homeostatic_arr)[:, 1]
+
+        kinetic_array = [[self.rxns.index(met), target] for met, target in kinetic_targets.items()]
+        kinetic_idx, kinetic_target = np.array(kinetic_array, dtype=np.int64)[:, 0], np.array(kinetic_array)[:, 1]
+
+        if binary_kinetic_targets:
+            binary_kinetic_idx = np.array([self.rxns.index(met) for met in binary_kinetic_targets.keys()], dtype=np.int64)
+        else :
+            binary_kinetic_idx = None
+
+        maintenance_idx = self.rxns.index("maintenance_reaction")  # TODO (Cyrus) - use name provided
+
+        # set up variables
+        v = cp.Variable(self.n_orig_rxns)
+        e = cp.Variable(self.n_exch_rxns)
+        dm = self.S_orig @ v + self.S_exch @ e
+        exch = self.S_exch @ e
+
+        constr = []
+        constr.append(dm[self.intermediates_idx] == 0)
+        constr.append(v[maintenance_idx] == maintenance_target)
+        # check if binary_kinetic_idx is empty
+        if binary_kinetic_idx is not None:
+            constr.append(v[binary_kinetic_idx] == 0)
+        # constr.append(dm[homeostatic_idx[]] == homeostatic_target)
+        constr.extend([v >= 0, v <= upper_flux_bound, e >= 0, e <= upper_flux_bound]) # TODO (Cyrus) - remove hard-coded bounds
+
+        loss = 0
+        loss += cp.norm1(dm[homeostatic_idx] - homeostatic_target)
+        loss += objective_weights['secretion'] * (cp.sum(e[self.secretion_idx]))
+        loss += objective_weights['efficiency'] * (cp.sum(v))
+        loss += objective_weights['kinetics'] * cp.norm1(v[kinetic_idx] - kinetic_target)
+
+        p = cp.Problem(
+            cp.Minimize(loss),
+            constr
+        )
+
+        p.solve(solver=cp.GLOP, verbose=False)
```

Comment:
Is there ever a case where we end up with an infeasible solution (not sure if that's the right term: I remember warning messages with the old metabolism saying it had to retry)?